lib/tls.js

```diff
@@ -143,9 +146,13 @@ exports.convertALPNProtocols = function convertALPNProtocols(protocols, out) {
   // If protocols is Array - translate it into buffer
   if (ArrayIsArray(protocols)) {
     out.ALPNProtocols = convertProtocols(protocols);
-  } else if (isArrayBufferView(protocols)) {
+  } else if (Buffer.isBuffer(protocols) || isUint8Array(protocols)) {
```

Comment:
Are `Buffer.isBuffer()` and `isUint8Array()` faster than `isArrayBufferView()`? If not we can remove this branch.

Reply:
The question here is, is `Buffer.from(buffer)` faster or better than `Buffer.from(arrayBuffer, byteOffset, length)`? Having `Buffer` or `Uint8Array` here is expected in most cases, otherwise we would got bugreports for this.
I don't really know but my guess is "yes, it should be faster" because `Buffer` tends to use shared pools, and by doing `buffer.slice()` we would copy the whole pool.

Reply:
> I don't really know but my guess is "yes, it should be faster" because `Buffer` tends to use shared pools, and by doing `buffer.slice()` we would copy the whole pool.

We should keep it as is when `protocols` is a `Buffer` but when it is a `Uint8Array` I'm not sure, AFAIK there is no pool.

Reply:
Anyway it is not very important, it would only slightly improve readability by removing the OR condition. Feel free to ignore.

Reply:
Hmm, tested it and
```mjs
const { Buffer } = require('node:buffer');
console.log(Buffer.from(new Uint8Array([1,2,3])).buffer.byteLength);
```
Shows 8192 😕

Reply:
In practice it would be

```js
const arr = new Uint8Array([1,2,3]);
Buffer.from(arr.slice(), arr.byteOffset, arr.byteLength);
```

and `arr.slice()` will not copy the whole pool as there is no pool?

Reply:
Isn't that the point of speedup? I mean, because of that in practice we *should* prefer doing `Buffer.from(arr)` directly so it will be able to reuse preallocated pool.

Reply:
```
$ node
Welcome to Node.js v18.2.0.
Type ".help" for more information.
> const arr = new Uint8Array([1,2,3]);
undefined
> arr.buffer.byteLength
3
> const buf = Buffer.from(arr.slice(), arr.byteOffset, arr.byteLength);
undefined
> buf.buffer.byteLength
8192
```

Reply:
```console
$ ./node
Welcome to Node.js v19.0.0-pre.
Type ".help" for more information.
> const arr = new Uint8Array([1,2,3]);
undefined
> const buf = Buffer.from(arr.buffer.slice(), arr.byteOffset, arr.byteLength);
undefined
> buf.buffer.byteLength
3
```
The one above uses `Uint8Array` directly rather than `ArrayBuffer` :)

Reply:
Oh right, sorry.

Reply:
> Isn't that the point of speedup? I mean, because of that in practice we should prefer doing Buffer.from(arr) directly so it will be able to reuse preallocated pool.

Yes, it makes sense.

Reply:
Actually, it was the right direction, applicable to the next branch. I forgot that `TypedArray.prototype.slice().buffer`, unlike `Buffer`'s version, does exactly what's needed there. Calling `.buffer.slice()` without arguments is in fact a huge mistake.
Unfortunately, `DataView` is not a `TypedArray` and supporting it requires slightly different approach.

Synthetic microbenchmarking shows ~50% speedup for `TypedArray` way over `DataView`, and ~700% for `Buffer|Uint8Array` way over `DataView` for me.
If readability suffered too much from that, last `else if (isArrayBufferView)` branch can be removed. Otherwise it seems to be the most correct workaround (at least until `Buffer.from(anyArrayBufferView)` is somehow supported 😄).